src/parse/asp/exec.go

```diff
@@ -96,6 +185,167 @@ func execGetCachedOutput(key execKey, args []string) (output string, found bool)
 	}
 }
 
+// execGitBranch returns the output of a git_branch() command.
+//
+// git_branch() returns the output of `git symbolic-ref -q --short HEAD`
+func execGitBranch(s *scope, args []pyObject) pyObject {
+	short := args[0].IsTruthy()
+
+	cmdIn := make([]pyObject, 3, 5)
+	cmdIn[0] = pyString("git")
+	cmdIn[1] = pyString("symbolic-ref")
+	cmdIn[2] = pyString("-q")
+	if short {
+		cmdIn = append(cmdIn, pyString("--short"))
+	}
+	cmdIn = append(cmdIn, pyString("HEAD"))
+
+	wantStdout := True
+	wantStderr := False
+	cacheOutput := True
+	keyExtra := pyString("")
+
+	execArgs := []pyObject{
+		pyList(cmdIn),
+		wantStdout,
+		wantStderr,
+		cacheOutput,
+		keyExtra,
+	}
+
+	return doExec(s, execArgs)
+}
+
+// execGitCommit returns the output of a git_commit() command.
+//
+// git_commit() returns the output of `git rev-parse --short HEAD`
+func execGitCommit(s *scope, args []pyObject) pyObject {
+	short := args[0].IsTruthy()
+
+	cmdIn := make([]pyObject, 2, 4)
+	cmdIn[0] = pyString("git")
+	cmdIn[1] = pyString("rev-parse")
+	if short {
+		cmdIn = append(cmdIn, pyString("--short"))
+	}
+	cmdIn = append(cmdIn, pyString("HEAD"))
+
+	wantStdout := True
+	wantStderr := False
+	cacheOutput := True
+	keyExtra := pyString("")
+
+	execArgs := []pyObject{
+		pyList(cmdIn),
+		wantStdout,
+		wantStderr,
+		cacheOutput,
+		keyExtra,
+	}
+
+	return doExec(s, execArgs)
+}
+
+// execGitShow returns the output of a git_show() command with a strict format.
+//
+// git_show() returns the output of `git show -s --format=%{fmt}`
+//
+// %ci == commit-date:
+//   `git show -s --format=%ci` = 2018-12-10 00:53:35 -0800
+func execGitShow(s *scope, args []pyObject) pyObject {
+	formatVerb := args[0].(pyString)
+	switch formatVerb {
```

Comment:
why do we need to check the format so closely? can we not just leave that up to git - or are you trying to filter out some of them?

Reply:
We're definitely trying to filter some of them out.  I didn't want to permit any of the ASCII control characters or audit for possible side effects from potential expansions, etc.  I'm okay relaxing this in the future, but I haven't dug into `git show`'s code to see what or how this could be abused.  For instance, I don't know what in `git show`, and by extension parts of `git config` let someone pass in a `PAGER` or some other value that would be interpreted as a shell command.

Reply:
ugh right, makes sense then :)

Reply:
I agree with restricting this.